MPU/src/driverioHW.cpp

```diff
@@ -0,0 +1,208 @@
+#include <nerduino.h>
+#include "driverioHW.h"
+
+/**************************************************************************************/
+/**
+ * @brief Button Class Implementation
+ */
+/**************************************************************************************/
+
+BUTTON::BUTTON(){}
+
+BUTTON::BUTTON(uint8_t pinNumber)
+{
+    pin = pinNumber;
+    pinMode(pin, INPUT_PULLUP);
+    debounce.cancelTimer();
+}
+
+BUTTON::~BUTTON(){}
+
+void BUTTON::checkButtonPin()
+{
+    if(digitalRead(pin) == HIGH && state == NOT_PRESSED)
+    {
+        state = DEBOUNCING;
+        debounce.startTimer(BUTTON_DEBOUNCE_TIME);
+    }
+    else if(digitalRead(pin) == LOW)
+    {
+        debounce.cancelTimer();
+        state = NOT_PRESSED;
+    }
+}
+
+bool BUTTON::isButtonPressed()
+{
+    if(state == PRESSED) return true;
+
+    /**
+     * Returns true if
+     * 1. The button has successfully passed the debounce period
+     * 2. The debounce timer has not been canceled/reset
+     * 3. The button is still being held
+     */
+    if(digitalRead(pin) && debounce.isTimerExpired() && !debounce.isTimerReset())
+    {
+        state = PRESSED;
+        return true;
+    }
+    return false;
+}
+
+bool BUTTON::isButtonPressed_Pulse()
+{
+    if(!readyToReadPulse) return false;
+
+    /**
+     * Returns true if
+     * 1. The button has successfully passed the debounce period
+     * 2. The debounce timer has not been canceled/reset
+     * 3. The button is still being held
+     * 
+     * AND
+     * 
+     * 4. The state of the button has not been read since the button has passed the debounce
+     */
+    if(digitalRead(pin) && debounce.isTimerExpired() && !debounce.isTimerReset())
+    {
+        state = PRESSED;
+        readyToReadPulse = false;
+        return true; 
+    }
+
+    return false;
+}
+
+/**************************************************************************************/
+/**
+ * @brief Speaker Class Implementation
+ */
+/**************************************************************************************/
+
+SPEAKER::SPEAKER(){}
+
+SPEAKER::SPEAKER(uint8_t pinNumber)
+{
+    pin = pinNumber;
+    pinMode(pin, OUTPUT);
+    writeSpeaker(LOW);
+    waitTime.cancelTimer();
+}
+
+SPEAKER::~SPEAKER(){}
+
+void SPEAKER::writeSpeaker(bool state)
+{
+    //We need to invert the logic of the speaker due to the hardware
+    digitalWrite(pin,!state);
+}
+
+void SPEAKER::playSpeaker()
+{
+    writeSpeaker(HIGH);
+    waitTime.startTimer(SPEAKER_DURATION);
+}
+
+void SPEAKER::attemptToStopSpeaker()
```

Comment:
This function is that check

Reply:
Does it have to be a function? Or if it has to be a function, should it just be standard across all the hardware classes and be like "updateButton", "updateSpeaker", "updateLED" etc

Reply:
Ig i'd have to think about how many need refreshes like this, but if it's the majority it may be nice to be able to always just call like button.update().

Reply:
Might be good to have a "updateDriverIO" function in the bigger wrapper class that we can slap all the "update" functions in